lib/connections/limiter.go

```diff
@@ -39,97 +40,167 @@ func newLimiter(deviceID protocol.DeviceID, cfg *config.Wrapper) *limiter {
 		read:                rate.NewLimiter(rate.Inf, limiterBurstSize),
 		myID:                deviceID,
 		mu:                  &sync.Mutex{},
-		deviceReadLimiters:  new(sync.Map),
-		deviceWriteLimiters: new(sync.Map),
+		deviceReadLimiters:  make(map[protocol.DeviceID]*rate.Limiter),
+		deviceWriteLimiters: make(map[protocol.DeviceID]*rate.Limiter),
+		deviceMapMutex:      &sync.RWMutex{},
 	}
 
 	// Get initial device configuration
 	devices := cfg.RawCopy().Devices
 	for _, value := range devices {
 		value.MaxRecvKbps = -1
 		value.MaxSendKbps = -1
+
+		// Keep read/write limiters for every connected device
+		l.deviceWriteLimiters[value.DeviceID] = rate.NewLimiter(rate.Inf, limiterBurstSize)
+		l.deviceReadLimiters[value.DeviceID] = rate.NewLimiter(rate.Inf, limiterBurstSize)
 	}
 	prev := config.Configuration{Options: config.OptionsConfiguration{MaxRecvKbps: -1, MaxSendKbps: -1}, Devices: devices}
 
-	// Keep read/write limiters for every connected device
-	l.rebuildMap(prev)
-
 	cfg.Subscribe(l)
 	l.CommitConfiguration(prev, cfg.RawCopy())
 	return l
 }
 
-// Create new maps with new limiters
-func (lim *limiter) rebuildMap(to config.Configuration) {
-	deviceWriteLimiters := new(sync.Map)
-	deviceReadLimiters := new(sync.Map)
+// This function sets limiters according to corresponding DeviceConfiguration
+func (lim *limiter) setLimitsForDevice(v config.DeviceConfiguration, readLimiter, writeLimiter *rate.Limiter) {
+	// The rate variables are in KiB/s in the config (despite the camel casing
+	// of the name). We multiply by 1024 to get bytes/s.
+	if v.MaxRecvKbps <= 0 {
+		readLimiter.SetLimit(rate.Inf)
+	} else {
+		readLimiter.SetLimit(1024 * rate.Limit(v.MaxRecvKbps))
+	}
+
+	if v.MaxSendKbps <= 0 {
+		writeLimiter.SetLimit(rate.Inf)
+	} else {
+		writeLimiter.SetLimit(1024 * rate.Limit(v.MaxSendKbps))
+	}
+}
+
+// This function handles removing, adding and updating of device limiters.
+// Pass pointer to avoid copying. Pointer already points to copy of configuration
+// so we don't have to worry about modifying real config.
+func (lim *limiter) processDevicesConfiguration(from, to *config.Configuration) {
 
-	// copy *limiter in case remote device is still connected, when remote device is added we create new read/write limiters
+	devicesToRemove := make(map[protocol.DeviceID]bool)
+	// Get all devices that were previously in map, these are candidates for removal
+	for _, v := range from.Devices {
+		if v.DeviceID != to.MyID {
+			devicesToRemove[v.DeviceID] = true
+		}
+	}
```

Comment:
Could be `map[protocol.DeviceID]struct{}` and then later instead of `= false` do `delete(devicesToRemove, v.DeviceID)`. Then you can just iterate over the final map in the end without checking a bool.